notes/plutus-core/cek-budgeting-profiling/CostingDetails.md

```diff
@@ -0,0 +1,240 @@
+execution is terminated if the limit is exceeded at any point during execution.
+
+The CEK machine (`Language.UntypedPlutusCore.Evaluation.Machine.Cek`) declares the following
+types
+
+```
+data ExBudgetCategory fun = BTyInst | BApply | BIWrap | BUnwrap | BVar | BBuiltin fun  | BAST
+type CekExBudgetState fun = ExBudgetState (ExBudgetCategory fun)
+type CekExTally       fun = ExTally       (ExBudgetCategory fun)
+```
+
+and an instance of `SpendBudget`:
+```
+instance (Eq fun, Hashable fun, ToExMemory term) =>
+            SpendBudget (CekCarryingM term uni fun) fun (ExBudgetCategory fun) term where
+    spendBudget key budget = do
+        modifying exBudgetStateTally
+                (<> (ExTally (singleton key budget)))
+        newBudget <- exBudgetStateBudget <%= (<> budget)
+        mode <- asks cekEnvBudgetMode
+        case mode of
+            Counting -> pure ()
+            Restricting resb ->
+                when (exceedsBudget resb newBudget) $
+                    throwingWithCause _EvaluationError
+                        (UserEvaluationError $ CekOutOfExError resb newBudget)
+                        Nothing  -- No value available for error
+```
+
+The machine runs in a monad containing a `CekBudgetState` object recording
+costs.  Note that `spendBudget` updates both `_exBudgetStateTally` and
+`_exBudgetStateBudget`: the `modifying` line updates the accumulated
+per-operation costs with the costs of the current operation, and the line
+`newBudget <- exBudgetStateBudget <%= (<> budget)` adds the cost of the current
+operation to the total consumed so far (and returns the updated result for
+further use).  Both fields are updated irrespective of
+whether we're in `Counting` mode or `Restricting` mode, which may explain why
+the figures in [CekProfiling.md](./CekProfiling.md) don't show much difference
+between execution times for the two modes.  To attempt to remedy this, it seems
+sensible to refactor `spendBudget` so that it's only updating the field relevant
+to the current mode:
+
+```
+    spendBudget key budget = do
+        mode <- asks cekEnvBudgetMode
+        case mode of
+            Counting -> modifying exBudgetStateTally (<> (ExTally (singleton key budget)))
+            Restricting resb -> do
+                newBudget <- exBudgetStateBudget <%= (<> budget)
+                when (exceedsBudget resb newBudget) $
+                    throwingWithCause _EvaluationError
+                        (UserEvaluationError $ CekOutOfExError resb newBudget)
+                        Nothing
+```
+
+This does reduce execution times noticeably in `Restricting` mode (see table
+later), but for some reason it reintroduces a memory leak in `Counting` mode:
+many of the `nofib` examples fail to complete because they consume large amounts
+of memory very quickly.  Changing the code to 
+
+```
+    spendBudget key budget = do
+        newBudget <- exBudgetStateBudget <%= (<> budget)
```

Comment:
This does not look like it's strict enough, however I guess your guess is probably correct: we disturb the record by only accessing `_exBudgetStateBudget` at which point all previous thunks get forced and then we create a new one by modifying that field with `(<> budget)`. And that last thunk will be forced at the next iteration.

Reply:
Yeah, I believe this is exactly what happens. If we update the code to the following:

```haskell
counterVar :: IORef Int
counterVar = unsafePerformIO $ newIORef 0
{-# NOINLINE counterVar #-}

instance (Eq fun, Hashable fun, ToExMemory term) =>
            SpendBudget (CekCarryingM term uni fun) fun (ExBudgetCategory fun) term where
    spendBudget key budget = do
        let !counter = unsafePerformIO $ readIORef counterVar
        let !_ = unsafePerformIO $ modifyIORef' counterVar succ
        newBudget <- trace (show counter ++ ":1") exBudgetStateBudget <%= (<> trace (show counter ++ ":2") budget)
        mode <- asks cekEnvBudgetMode
        case mode of
            Counting -> modifying (trace (show counter ++ ":3") exBudgetStateTally) (<> trace (show counter ++ ":4") (ExTally (singleton key budget)))
            Restricting -> <...>
```

then this gives us the following output: 

```
0:1
1:1
0:3
0:2
0:4
2:1
1:3
1:2
1:4
3:1
2:3
2:2
2:4
<...>
```

where a number before `:` is an iteration counter and a number after `:` indicates a particular place in the code. The beginning of this sequence is

```
0:1
1:1
```

which demonstrates that `<%=` forces its first argument, but does not actually update the record (that becomes a thunk). `modifying` does not even force its first argument, let alone update the record. And so at the zeroth iteration we've only forced the record by accessing it in `<%=` and constructed a bunch of thunks. At the first iteration we access the record again and this causes all the thunks created previously to be forced as well, giving us

```
0:3
0:2
0:4
```

And then the process continues: accessing the record via `<%=` forces thunks created at the previous iteration.

So basically `<%=` and `modifying` both modify the state, but have different strictness properties.